http/tus.go

```
@@ -0,0 +1,247 @@
+	if !ok {
+		log.Printf("Creating tus handler for user %s\n", d.user.Username)
+		basePath, err := getBasePathFromRequest(r, th.apiPath)
+		if err != nil {
+			return nil, err
+		}
+		handler = th.createTusHandler(d, basePath.String())
+		th.handlers[d.user.ID] = handler
+	}
+
+	return handler, nil
+}
+
+func (th tusHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	code, err := withUser(func(w http.ResponseWriter, r *http.Request, d *data) (int, error) {
+		// Create a new tus handler for current user if it doesn't exist yet
+		// Use a mutex to make sure only one tus handler is created for each user
+		mutex.Lock()
+		handler, err := th.getOrCreateTusHandler(d, r)
+		mutex.Unlock()
+
+		if err != nil {
+			return 400, err
+		}
+
+		// Create upload directory for each request
+		uploadDir := filepath.Join(d.user.FullPath("/"), ".tmp_upload")
+		if err := os.MkdirAll(uploadDir, os.ModePerm); err != nil {
+			return http.StatusInternalServerError, err
+		}
+
+		switch r.Method {
+		case "POST":
+			handler.PostFile(w, r)
+		case "HEAD":
+			handler.HeadFile(w, r)
+		case "PATCH":
+			handler.PatchFile(w, r)
+		default:
+			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
+		}
+
+		// Isn't used
+		return 201, nil
+	})(w, r, &data{
+		store:    th.store,
+		settings: th.settings,
+		server:   th.server,
+	})
+
+	if err != nil {
+		http.Error(w, err.Error(), code)
+	} else if code >= 400 {
+		http.Error(w, "", code)
+	}
+}
+
+func (th tusHandler) createTusHandler(d *data, basePath string) *tusd.UnroutedHandler {
+	uploadDir := filepath.Join(d.user.FullPath("/"), th.uploadDirName)
+	tusStore := filestore.FileStore{
+		Path: uploadDir,
+	}
+	composer := tusd.NewStoreComposer()
+	tusStore.UseIn(composer)
+
+	handler, err := tusd.NewUnroutedHandler(tusd.Config{
+		BasePath:              basePath,
+		StoreComposer:         composer,
+		NotifyCompleteUploads: true,
+	})
+	if err != nil {
+		panic(fmt.Errorf("Unable to create handler: %s", err))
```

Why not return this error?